File: apps/visualizer/backend/api/scores.py
"""Read-only REST API for persisted ``image_scores`` rows (library DB)."""

from typing import Any

from flask import Blueprint, jsonify, request
from spectree import Response
from utils.db import with_db
from utils.responses import error_bad_request

from api.openapi import spec
from api.schemas.jobs import ErrorBody
from api.schemas.scores import ScoresCurrentResponse, ScoresHistoryResponse
from utils.perspective_slug import PERSPECTIVE_SLUG_RE as _SLUG_RE
from lightroom_tagger.core.database import (
    get_current_scores_for_image,
    list_score_history_for_perspective,
)

bp = Blueprint("scores", __name__)


def _normalize_score_row(row: dict) -> dict:
    out = dict(row)
    out["is_current"] = bool(out.get("is_current"))
    out["repaired_from_malformed"] = bool(out.get("repaired_from_malformed"))
    out["not_attempted"] = bool(out.get("not_attempted"))
    return out


def _image_type_from_request() -> tuple[str | None, Any]:
    raw = (request.args.get("image_type") or "catalog").strip().lower()
    if raw not in ("catalog", "instagram"):
        return None, error_bad_request("image_type must be catalog or instagram")
    return raw, None


@bp.route("/<path:image_key>/history", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=ScoresHistoryResponse, HTTP_400=ErrorBody),
    tags=['scores'],
)
def get_score_history(db, image_key: str):
    image_type, err = _image_type_from_request()
    if err is not None:
        return err
    assert image_type is not None

    perspective_slug = (request.args.get("perspective_slug") or "").strip()
    if not perspective_slug:
        return error_bad_request("perspective_slug is required")
    if not _SLUG_RE.match(perspective_slug):
        return error_bad_request("invalid perspective_slug")

    rows = list_score_history_for_perspective(db, image_key, image_type, perspective_slug)
    return jsonify(
        {
            "image_key": image_key,
            "image_type": image_type,
            "perspective_slug": perspective_slug,
            "history": [_normalize_score_row(r) for r in rows],
        }
    )


@bp.route("/<path:image_key>", methods=["GET"])
@with_db
@spec.validate(
    resp=Response(HTTP_200=ScoresCurrentResponse, HTTP_400=ErrorBody),
    tags=['scores'],
)
def get_current_scores(db, image_key: str):
    image_type, err = _image_type_from_request()
    if err is not None:
        return err
    assert image_type is not None

    rows = get_current_scores_for_image(db, image_key, image_type)
    return jsonify(
        {
            "image_key": image_key,
            "image_type": image_type,
            "current": [_normalize_score_row(r) for r in rows],
        }
    )

File: apps/visualizer/backend-ts/src/api/scores.ts
/**
 * Read-only REST API for persisted `image_scores` rows (library DB).
 */
import { createRoute, z } from '@hono/zod-openapi';
import {
  getCurrentScoresForImage,
  listScoreHistoryForPerspective,
  type ImageScoreRow as DbScoreRow,
} from '../db/library/scores.js';
import { libraryDb, type LibraryEnv } from '../db/library/with-db.js';
import { isValidPerspectiveSlug } from '../utils/perspective-slug.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  ImageScoreRow,
  ScoresCurrentResponse,
  ScoresHistoryResponse,
} from './schemas/scores.js';

export const scoresRoutes = createOpenApiApp<LibraryEnv>();

// Every scores route reads the library DB; open it once per request here.
scoresRoutes.use('*', libraryDb());

/** SQLite stores these as 0/1; the API contract is boolean. */
function normalizeScoreRow(row: DbScoreRow): z.infer<typeof ImageScoreRow> {
  return {
    ...row,
    is_current: Boolean(row.is_current),
    repaired_from_malformed: Boolean(row.repaired_from_malformed),
    not_attempted: Boolean(row.not_attempted),
  };
}

/**
 * `image_type` is accepted but only `catalog` is valid — the Instagram scope was
 * removed (#218). Kept as a parameter so existing callers get a clear 400 rather
 * than silently different data.
 */
function readImageType(raw: string | undefined): { ok: true; value: 'catalog' } | { ok: false } {
  const value = (raw ?? 'catalog').trim().toLowerCase();
  return value === 'catalog' ? { ok: true, value } : { ok: false };
}

/**
 * Deliberate narrowing: Flask routed these with the `path:` converter, which also
 * matches slashes. A plain Hono parameter does not.
 *
 * Verified against the catalog: 0 of 43,794 image keys contain a slash, and none in
 * `image_scores` either — keys use only alphanumerics, `-`, `_`, space and
 * parentheses. A plain parameter therefore covers every real key, and avoids the
 * routing hazard a greedy `{.+}` introduces, where `/history` would be swallowed
 * into the key unless route registration order is exactly right.
 *
 * If slashed keys ever become possible, this needs `{.+}` plus an ordering test.
 */
const historyRoute = createRoute({
  method: 'get',
  path: '/{image_key}/history',
  tags: ['scores'],
  request: { params: z.object({ image_key: z.string() }) },
  responses: withValidationError({
    200: { description: 'Score history for one perspective', content: jsonBody(ScoresHistoryResponse) },
    400: { description: 'Invalid request', content: jsonBody(ErrorBody) },
  }),
});

scoresRoutes.openapi(historyRoute, (c) => {
  const imageKey = c.req.param('image_key');
  const imageType = readImageType(c.req.query('image_type'));
  if (!imageType.ok) return c.json({ error: 'image_type must be catalog' }, 400);

  const perspectiveSlug = (c.req.query('perspective_slug') ?? '').trim();
  if (!perspectiveSlug) return c.json({ error: 'perspective_slug is required' }, 400);
  if (!isValidPerspectiveSlug(perspectiveSlug)) {
    return c.json({ error: 'invalid perspective_slug' }, 400);
  }

  const rows = listScoreHistoryForPerspective(
    c.get('libraryDb'),
    imageKey,
    imageType.value,
    perspectiveSlug,
  );
  return c.json(
    {
      image_key: imageKey,
      image_type: imageType.value,
      perspective_slug: perspectiveSlug,
      history: rows.map(normalizeScoreRow),
    },
    200,
  );
});

const currentRoute = createRoute({
  method: 'get',
  path: '/{image_key}',
  tags: ['scores'],
  request: { params: z.object({ image_key: z.string() }) },
  responses: withValidationError({
    200: { description: 'Current scores for an image', content: jsonBody(ScoresCurrentResponse) },
    400: { description: 'Invalid request', content: jsonBody(ErrorBody) },
  }),
});

scoresRoutes.openapi(currentRoute, (c) => {
  const imageKey = c.req.param('image_key');
  const imageType = readImageType(c.req.query('image_type'));
  if (!imageType.ok) return c.json({ error: 'image_type must be catalog' }, 400);

  const rows = getCurrentScoresForImage(c.get('libraryDb'), imageKey, imageType.value);
  return c.json(
    {
      image_key: imageKey,
      image_type: imageType.value,
      current: rows.map(normalizeScoreRow),
    },
    200,
  );
});
